### ISSUE TITLE
type checking for attributes

### ISSUE THREAD

**Issue by glyph** [Shoobx/mypy-zope#20]
This example doesn't complain:

```
from zope.interface import Interface, Attribute, implementer

class IWhatever(Interface):
    stuff: str = Attribute("name")


@implementer(IWhatever)
class WhateverBroken(object):
    stuff: int
```

even though the type of `stuff` doesn't match.

This closed PR, from 2017, suggests that this syntax was intended to work one day:

https://github.com/zopefoundation/zope.interface/issues/98

I'm guessing this is what the README is already talking about, when it says "Interface compatibility checker will not type-check non-method attributes"?

**Comment by kedder** [comment]
@glyph, yes, that README sentence refers to this issue. This feature was supported in the earlier version, but since interface checking was reimplemented, I had to drop it. The reason is documentde in `_report_implementation_problems`:

https://github.com/Shoobx/mypy-zope/blob/ae7ed4bebe15a37976cf49dbbbf31d317bc7f813/src/mypy_zope/plugin.py#L553-L559

**Comment by kedder** [comment]
If I remember correctly, the issue is boils down to the type difference between class and instance attributes, e.g.

```python

class Descriptor:
    __get__(self, obj) -> int:
        return 0

class IWhatever(Interface):
    stuff: int = Attribute("name")

@implementer(IWhatever)
class Whatever:
    stuff = Descriptor()
```

The type of `Whatever.stuff` will be `Descriptor`, but type of `Whatever().stuff` will be `int`. `IWhatever.stuff` is int, but `Whatever.stuff` is `Descriptor`, which would be reported as mismatch.

**Comment by glyph** [comment]
@kedder Thanks for the explanation. I went looking for how this was handled by `Protocol` to see how it could work, but it looks like `Protocol`s have this same issue!  I added `__set__` and `__delete__` methods to make sure that it wasn't an issue with the property declared as mutable being incomplete, but this still results in an error:

```
from typing import Protocol

class Descriptor:
    def __get__(self, obj: object) -> int:
        return 0

    def __set__(self, obj: object, value: int) -> None:
        return None

    def __delete__(self, obj: object) -> None:
        return None

class Whatever:
    stuff = Descriptor()


class WhateverProto(Protocol):
    stuff: int


whatever: WhateverProto = Whatever()
```

**Comment by glyph** [comment]
Looks like it's a known issue: https://github.com/python/mypy/issues/5481

**Comment by glyph** [comment]
Would it be possible to make this a configuration option?  There are some hacks which would make this possible (like what `attrs`'s stubs do with `Factory`)

**Comment by kedder** [comment]
We can definiely make it an option. 

@glyph, can you try uncommenting these lines and check if that actually works for your project?

https://github.com/Shoobx/mypy-zope/blob/ae7ed4bebe15a37976cf49dbbbf31d317bc7f813/src/mypy_zope/plugin.py#L561-L567

**Comment by glyph** [comment]
Gotcha. There may be … significant latency on this experiment, since we're burning down LOTS of unrelated mypy errors on Twisted right now, so seeing how this works in practice might take a while.  But @adiroiban @rodrigc might be interested to know that this is possible.
